### ISSUE TITLE
 Couldn't find an alternative telinit implementation to spawn.

### ISSUE THREAD

**Issue by inspired-geek** [gluster/gluster-kubernetes#298]
Hello! I'm having issue with freshly installed k8s cluster

```
# kubectl --version
Kubernetes v1.7.3
```
 on 4 servers with latest CentOS

```
# cat /etc/redhat-release
CentOS Linux release 7.3.1611 (Core)
```

with docker CE 17.05-0

```
# docker -v
Docker version 17.05.0-ce, build 89658be
```

After pull repository I've set up topology.json and run `.\dk-deploy -g` in `deploy` folder.

```
# ./gk-deploy -g
Welcome to the deployment tool for GlusterFS on Kubernetes and OpenShift.

Before getting started, this script has some requirements of the execution
environment and of the container platform that you should verify.

The client machine that will run this script must have:
 * Administrative access to an existing Kubernetes or OpenShift cluster
 * Access to a python interpreter 'python'
 * Access to the heketi client 'heketi-cli'

Each of the nodes that will host GlusterFS must also have appropriate firewall
rules for the required GlusterFS ports:
 * 2222  - sshd (if running GlusterFS in a pod)
 * 24007 - GlusterFS Daemon
 * 24008 - GlusterFS Management
 * 49152 to 49251 - Each brick for every volume on the host requires its own
   port. For every new brick, one new port will be used starting at 49152. We
   recommend a default range of 49152-49251 on each host, though you can adjust
   this to fit your needs.

For systems with SELinux, the following settings need to be considered:
 * virt_sandbox_use_fusefs should be enabled on each node to allow writing to
   remote GlusterFS volumes

In addition, for an OpenShift deployment you must:
 * Have 'cluster_admin' role on the administrative account doing the deployment
 * Add the 'default' and 'router' Service Accounts to the 'privileged' SCC
 * Have a router deployed that is configured to allow apps to access services
   running in the cluster

Do you wish to proceed with deployment?

[Y]es, [N]o? [Default: Y]:
Using Kubernetes CLI.
NAME      STATUS    AGE
default   Active    15h
Using namespace "default".
Checking that heketi pod is not running ... OK
serviceaccount "heketi-service-account" created
clusterrolebinding "heketi-sa-view" created
node "kub66" labeled
node "kub67" labeled
node "kub68" labeled
node "kub69" labeled
daemonset "glusterfs" created
Waiting for GlusterFS pods to start ...
```
and stuck with it

In k8s dashboard I can see deployed daemon set
![image](https://user-images.githubusercontent.com/1792014/29314298-075bd4f6-81c6-11e7-8ed6-5ee000732428.png)

and failed pods
![image](https://user-images.githubusercontent.com/1792014/29314309-1761b6e0-81c6-11e7-894c-16bf732969b2.png)

And one line in each container log
![image](https://user-images.githubusercontent.com/1792014/29314320-2b4b9752-81c6-11e7-8e26-fae19825db45.png)

```
Couldn't find an alternative telinit implementation to spawn.
```

As far as I can understand it is about systemd permissions. There is an issue in kubernetes itself about it (https://github.com/kubernetes/kubernetes/issues/48937)

Is there any workaround of fix for this problem?

**Comment by jarrpa** [comment]
This is the first we hear of this. At least it confirms some suspicions I had... :)

It doesn't look like this has any real workaround we can implement at this time. For now all I found there was this comment: https://github.com/kubernetes/kubernetes/issues/48937#issuecomment-320956823

**Comment by hpgood** [comment]
I have the same problem.

**Comment by inspired-geek** [comment]
I've dowgraded docker on k8s hosts to 1.12 and it worked out.
But I don't think it's a good solution.

Is it possible to get rid of systemd and split required services in different containers like it was supposed in https://github.com/kubernetes/kubernetes/issues/48937#issuecomment-321243669 ?

**Comment by hpgood** [comment]
with the suggestion of(kubernetes/kubernetes#48937 (comment) ),I pass this bug,by using the file: (deploy/kube-templates/glusterfs-daemonset.yaml)

1.https://github.com/hpgood/heketi/blob/master/fix/glusterfs-daemonset.yaml
and use the new heketi :

2.https://github.com/hpgood/heketi/  
https://github.com/hpgood/heketi/commit/6eebb3ff9f2f5916f6e1e22b9246dac5b1c12092

**Comment by jarrpa** [comment]
@hpgood Good enough for now! Can you submit PRs to both gluster-kubernetes and heketi explaining why they're needed?

**Comment by hpgood** [comment]
@jarrpa I found that complime heketi is hard work( it can not support go source of kubernetes 1.7).
so I use anothor way to do step 2:
use a custom fix gluster-centos images by:
https://github.com/hpgood/heketi/tree/master/fix

now I can work with glusterfs in kubernetes 1.7.2 ,docker 17.06 ,rancher 1.6.6

**Comment by jarrpa** [comment]
@hpgood What's the diff between our current glusterfs-daemonset.yaml and yours?

**Comment by hpgood** [comment]
1.add 
```
env:
        - name: SYSTEMD_IGNORE_CHROOT
          value: "1"
  and
command:
        - /usr/lib/systemd/systemd
        - --system
```
 reason: https://github.com/kubernetes/kubernetes/issues/48937   

2.add
```
readinessProbe:
       tcpSocket:
            port: 24007
  livenessProbe:
       tcpSocket:
            port: 24007
```
reson:   in some case(first init time) ,the check script `systemctl status glusterd` cause a error(

> Failed to get D-Bus connection

). but it is ok after a short wait .

**Comment by jarrpa** [comment]
@hpgood Cool! I'll have to check this out myself. :)

@humblec Can you have a look at this and see if this would break anything?

**Comment by humblec** [comment]
@jarrpa Sure.
@hpgood `failed to get DBus connection` mostly occurs when the container is not running with privilege. 

@jarrpa I would like to wait for kube issue discussion to reach to consensus, before merging this PRs. I would request to hold PR merge till we conclude.

**Comment by hpgood** [comment]
![20170821](https://user-images.githubusercontent.com/3827559/29505489-42f43262-8679-11e7-8f87-71a2b644b179.jpg)

with the fix,after a short wait,it will trune to ok.

But for now ,maybe this fix is only choose to work with kubernetes 1.7.2 and docker-ce 17.06,my project cannot wait the next version of kubernetes. and I don't want to use the old version of docker and kubernetes. when the next release version of kubernetes  fix it in next time(many  months ),I think i will make a upgrade job.

**Comment by hpgood** [comment]
@jarrpa  I am testing this week.
I found that two service fail, and others is run success.

1. gluster-setup.service fail
2. rpcbind.socket fail


● gluster-setup.service - Configuring GlusterFS in container
   Loaded: loaded (/etc/systemd/system/gluster-setup.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Thu 2017-08-17 07:09:49 UTC; 4 days ago
  Process: 33 ExecStart=/usr/sbin/gluster-setup.sh (code=exited, status=1/FAILURE)
 Main PID: 33 (code=exited, status=1/FAILURE)

Aug 17 07:09:49 node-002 gluster-setup.sh[33]: ls: cannot access /var/lib/heketi/fstab: No such file or directory
Aug 17 07:09:49 node-002 gluster-setup.sh[33]: heketi-fstab not found
Aug 17 07:09:49 node-002 gluster-setup.sh[33]: /etc/glusterfs is not empty
Aug 17 07:09:49 node-002 gluster-setup.sh[33]: /var/log/glusterfs is not empty
Aug 17 07:09:49 node-002 gluster-setup.sh[33]: /var/lib/glusterd is not empty



● rpcbind.socket - RPCbind Server Activation Socket
   Loaded: loaded (/usr/lib/systemd/system/rpcbind.socket; enabled; vendor preset: enabled)
   Active: failed (Result: resources)
   Listen: /var/run/rpcbind.sock (Stream)
           0.0.0.0:111 (Stream)

**Comment by hpgood** [comment]
but for now,I found that: /var/lib/heketi/fstab is exist!

 cat /var/lib/heketi/fstab
/dev/mapper/vg_b693fcb53893474e9671ff3657568c8b-brick_b0e8aa6219540873a8db3609e8df03e0 /var/lib/heketi/mounts/vg_b693fcb53893474e9671ff3657568c8b/brick_b0e8aa6219540873a8db3609e8df03e0 xfs rw,inode64,noatime,nouuid 1 2
/dev/mapper/vg_b693fcb53893474e9671ff3657568c8b-brick_d7ba685e278cd9ea3b6814ec15485292 /var/lib/heketi/mounts/vg_b693fcb53893474e9671ff3657568c8b/brick_d7ba685e278cd9ea3b6814ec15485292 xfs rw,inode64,noatime,nouuid 1 2
/dev/mapper/vg_b693fcb53893474e9671ff3657568c8b-brick_c0446a84b130ac5104fb341e49df2209 /var/lib/heketi/mounts/vg_b693fcb53893474e9671ff3657568c8b/brick_c0446a84b130ac5104fb341e49df2209 xfs rw,inode64,noatime,nouuid 1 2
/dev/mapper/vg_b693fcb53893474e9671ff3657568c8b-brick_002bf6f695993f02a32d70cac6c277c6 /var/lib/heketi/mounts/vg_b693fcb53893474e9671ff3657568c8b/brick_002bf6f695993f02a32d70cac6c277c6 xfs rw,inode64,noatime,nouuid 1 2

**Comment by jarrpa** [comment]
Does the problem resolve itself if you revert to an older version of either Kube or Docker and use the original Daemonset and image?

**Comment by hpgood** [comment]
> I've dowgraded docker on k8s hosts to 1.12 and it worked out.
> But I don't think it's a good solution.
> 
> Is it possible to get rid of systemd and split required services in different containers like it was supposed in kubernetes/kubernetes#48937 (comment) ? 

inspired-geek had try it,I do not try revert the older version of them.

**Comment by jarrpa** [comment]
All right... let us know if you can manage to figure out why the services are failing. I'll try to track down someone who knows systemd in the meantime.

**Comment by hpgood** [comment]
> [root@node-002 /]# cat /usr/sbin/gluster-setup.sh 
> #!/bin/bash
> 
> ###
> # Description: Script to move the glusterfs initial setup to bind mounted directories of Atomic Host.
> # Copyright (c) 2016 Red Hat, Inc. <http://www.redhat.com>
> #
> # This file is part of GlusterFS.
> #
> # This file is licensed to you under your choice of the GNU Lesser
> # General Public License, version 3 or any later version (LGPLv3 or
> # later), or the GNU General Public License, version 2 (GPLv2), in all
> # cases as published by the Free Software Foundation.
> ###
> 
> main () {
>   if test "$(ls /var/lib/heketi/fstab)"
>   then
>         mount -a --fstab /var/lib/heketi/fstab
>         if [ $? -eq 1 ]
>         then
>               echo "mount failed"
>               exit 1
>         fi
>         echo "Mount Successful"
>   else
>         echo "heketi-fstab not found"
>   fi
>   DIR_1="/etc/glusterfs"
>   DIR_2="/var/log/glusterfs"
>   DIR_3="/var/lib/glusterd"
>   var=0
>   for i in $DIR_1 $DIR_2 $DIR_3
>   do
>     if test "$(ls $i)"
>     then
>       echo "$i is not empty"
>       var=$((var+1))
>     fi
>   done
> 
>   if [ $var -eq 3 ]
>   then
>         exit 1
>   fi
>   
>   cp -r /etc/glusterfs_bkp/* /etc/glusterfs
>   if [ $? -eq 1 ]
>   then
>         echo "Failed to copy $DIR_1" 
>         exit 1
>   fi
> 
>   cp -r /var/log/glusterfs_bkp/* /var/log/glusterfs
>   if [ $? -eq 1 ]
>   then
>         echo "Failed to copy $DIR_2"
>         exit 1
>   fi
> 
>   cp -r /var/lib/glusterd_bkp/* /var/lib/glusterd
>   if [ $? -eq 1 ]
>   then
>         echo "Failed to copy $DIR_3"
>         exit 1
>   fi
>  
>   echo "Script Ran Successfully"
> }
> main
I think the reson is : I do not clean those dirs which created in the retry step before.
DIR_1="/etc/glusterfs"
DIR_2="/var/log/glusterfs"
DIR_3="/var/lib/glusterd"

**Comment by inspired-geek** [comment]
> inspired-geek had try it,I do not try revert the older version of them.

Yes, I've tried to revert back to 1.12 and it worked with kube 1.7.2 (not 1.7.3 due to https://github.com/kubernetes/kubernetes/issues/50463).
 But eventually I've chosen installing glusterfs-server directly on hosts. I haven't managed to setup glusterfs cluster with gk-deploy script.

**Comment by mafzst** [comment]
@inspired-geek Working for me with docker 1.12.6 and kube 1.7.4

Downgrading docker was the solution for me

**Comment by YorikSar** [comment]
FIY You can add `--docker-disable-shared-pid` flag to kubelet to prevent this behavior. This worked for me.

**Comment by jarrpa** [comment]
If anyone here is still interested in experimenting, I uploaded experimental Docker images of GlusterFS without systemd:  This image includes a dummy `systemctl` that just returns true, to satisfy heketi. ;)
* https://hub.docker.com/r/jarrpa/gluster-fedora-minimal/tags/ - based on Fedora
  * `latest` is the one you probably want
  * `block` includes gluster-blockd, which is still under development
* https://hub.docker.com/r/jarrpa/gluster-ubuntu-minimal/tags/ - based on ubuntu-slim
  * `latest` again here

It also must be paired with a new daemonset definition, which is based on @hpgood's work: https://github.com/jarrpa/gluster-kubernetes/blob/jarrpa-dev/deploy/kube-templates/glusterfs-daemonset.yaml

Please reply if you get a chance to play with it!

**Comment by panigrahis** [comment]
@jarrpa I am facing the exact same issue.
Just to confirm I need to do two things:
a) Use  @hpgood glusterfs-daemonset.yaml
b) Replace image gluster/gluster-centos:latest with jarrpa/gluster-ubuntu-minimal:latest
Please let me know if my understanding is correct.

**Comment by jarrpa** [comment]
@panigrahis Yeah. I think you can also just use the yaml file I linked, as it has that change already.

**Comment by panigrahis** [comment]
@jarrpa got over the earlier issue.
Now am able to access logs but the deployment is still failing with following error:
root@GlusterServer01:~# kubectl logs glusterfs-4dl01
/etc/glusterfs is not empty
/var/lib/glusterd is not empty
ls: cannot access '/var/lib/heketi/fstab': No such file or directory
heketi-fstab not found
rpcinfo: can't contact rpcbind: RPC: (unknown error code)
Starting rpcbind
[2017-08-30 14:44:15.480890] I [MSGID: 100030] [glusterfsd.c:2475:main] 0-/usr/sbin/glusterd: Started running /usr/sbin/glusterd version 3.10.5 (args: /usr/sbin/glusterd -p /var/run/glusterd.pid --log-level INFO --no-daemon --debug)
[2017-08-30 14:44:15.481002] D [MSGID: 0] [glusterfsd.c:2088:glusterfs_pidfile_update] 0-glusterfsd: pidfile /var/run/glusterd.pid updated with pid 115
[2017-08-30 14:44:15.481018] D [logging.c:1790:__gf_log_inject_timer_event] 0-logging-infra: Starting timer now. Timeout = 120, current buf size = 5
[2017-08-30 14:44:15.481423] D [MSGID: 0] [glusterfsd.c:657:get_volfp] 0-glusterfsd: loading volume file /etc/glusterfs/glusterd.vol
[2017-08-30 14:44:15.486613] I [MSGID: 106478] [glusterd.c:1449:init] 0-management: Maximum allowed open file descriptors set to 65536
[2017-08-30 14:44:15.486658] I [MSGID: 106479] [glusterd.c:1496:init] 0-management: Using /var/lib/glusterd as working directory
[2017-08-30 14:44:15.486735] D [MSGID: 0] [glusterd.c:477:glusterd_rpcsvc_options_build] 0-glusterd: listen-backlog value: 128
[2017-08-30 14:44:15.486773] D [rpcsvc.c:2339:rpcsvc_init] 0-rpc-service: RPC service inited.
[2017-08-30 14:44:15.486786] D [rpcsvc.c:1889:rpcsvc_program_register] 0-rpc-service: New program registered: GF-DUMP, Num: 123451501, Ver: 1, Port: 0
[2017-08-30 14:44:15.486804] D [rpc-transport.c:279:rpc_transport_load] 0-rpc-transport: attempt to load file /usr/lib/x86_64-linux-gnu/glusterfs/3.10.5/rpc-transport/socket.so
[2017-08-30 14:44:15.489388] D [socket.c:4104:socket_init] 0-socket.management: Configued transport.tcp-user-timeout=0
[2017-08-30 14:44:15.489424] D [socket.c:4122:socket_init] 0-socket.management: Reconfigued transport.keepalivecnt=9
[2017-08-30 14:44:15.489434] D [socket.c:4205:socket_init] 0-socket.management: SSL support on the I/O path is NOT enabled
[2017-08-30 14:44:15.489438] D [socket.c:4208:socket_init] 0-socket.management: SSL support for glusterd is NOT enabled
[2017-08-30 14:44:15.489442] D [socket.c:4225:socket_init] 0-socket.management: using system polling thread
[2017-08-30 14:44:15.489452] D [name.c:584:server_fill_address_family] 0-socket.management: option address-family not specified, defaulting to inet
[2017-08-30 14:44:15.489503] E [socket.c:796:__socket_server_bind] 0-socket.management: binding to  failed: Address already in use
[2017-08-30 14:44:15.489515] E [socket.c:799:__socket_server_bind] 0-socket.management: Port is already in use
[2017-08-30 14:44:15.489534] W [rpcsvc.c:1668:rpcsvc_create_listener] 0-rpc-service: listening on transport failed
[2017-08-30 14:44:15.489548] E [MSGID: 106243] [glusterd.c:1713:init] 0-management: creation of listener failed
[2017-08-30 14:44:15.489565] E [MSGID: 101019] [xlator.c:503:xlator_init] 0-management: Initialization of volume 'management' failed, review your volfile again
[2017-08-30 14:44:15.489573] E [MSGID: 101066] [graph.c:325:glusterfs_graph_init] 0-management: initializing translator failed
[2017-08-30 14:44:15.489586] E [MSGID: 101176] [graph.c:681:glusterfs_graph_activate] 0-graph: init failed
[2017-08-30 14:44:15.489908] D [logging.c:1764:gf_log_flush_extra_msgs] 0-logging-infra: Log buffer size reduced. About to flush 5 extra log messages
[2017-08-30 14:44:15.489931] D [logging.c:1767:gf_log_flush_extra_msgs] 0-logging-infra: Just flushed 5 extra log messages
[2017-08-30 14:44:15.490003] W [glusterfsd.c:1332:cleanup_and_exit] (-->/usr/sbin/glusterd(glusterfs_volumes_init+0xe5) [0x55cca07f6235] -->/usr/sbin/glusterd(glusterfs_process_volfp+0x1b7) [0x55cca07f60f7] -->/usr/sbin/glusterd(cleanup_and_exit+0x54) [0x55cca07f5684] ) 0-: received signum (1), shutting down
[2017-08-30 14:44:15.490025] D [glusterfsd-mgmt.c:2613:glusterfs_mgmt_pmap_signout] 0-fsd-mgmt: portmapper signout arguments not given

**Comment by jarrpa** [comment]
@panigrahis Delete the contents of '/var/lib/glusterd' and '/etc/glusterfs' on all nodes

**Comment by panigrahis** [comment]
@jarrpa do i have to stop the glusterfs-server service running on these nodes?

**Comment by jarrpa** [comment]
@panigrahis Yes. Destroy all Gluster-related resources (I think just deleting the daemonset gets the pods, too), and just to be even safer do:
```
rm -rf /var/lib/heketi /etc/glusterfs /var/lib/glusterd /var/log/glusterfs
```

**Comment by panigrahis** [comment]
@jarrpa  that helped all went through fine but am getting this error now:
Creating initial resources ... serviceaccount "heketi-service-account" created
clusterrolebinding "heketi-sa-view" created
clusterrolebinding "heketi-sa-view" labeled
OK
node "glusterserver01" labeled
node "glusterserver02" labeled
node "glusterserver03" labeled
daemonset "glusterfs" created
Waiting for GlusterFS pods to start ... OK
secret "heketi-config-secret" created
secret "heketi-config-secret" labeled
service "deploy-heketi" created
deployment "deploy-heketi" created
Waiting for deploy-heketi pod to start ... OK
Creating cluster ... ID: 3db5c14d9b9553e111cd050c77524b31
Creating node glusterserver01 ... ID: 1b69778aec9cbb6e6fc268c01c22c0c1
Adding device /dev/sdb ... OK
Adding device /dev/sdc ... OK
Adding device /dev/sdd ... OK
Adding device /dev/sde ... OK
Creating node glusterserver02 ... ID: 3cfab510c651dc1cfdec60b7784291d2
Adding device /dev/sdb ... OK
Adding device /dev/sdc ... OK
Adding device /dev/sdd ... OK
Adding device /dev/sde ... OK
Creating node glusterserver03 ... ID: 09f81225b52c659af5e5bf7b2af57cc8
Adding device /dev/sdb ... OK
Adding device /dev/sdc ... OK
Adding device /dev/sdd ... OK
Adding device /dev/sde ... OK
heketi topology loaded.
Error: Unable to execute command on glusterfs-5cxp0:   /sbin/modprobe: execvp failed: No such file or directory
/sbin/modprobe failed: 2
Cannot read thin-pool target version.
thin: Required device-mapper target(s) not detected in your kernel.
Run `lvcreate --help' for more information.
Failed on setup openshift heketi storage
This may indicate that the storage must be wiped and the GlusterFS nodes must be reset.

I know its so near and it failed... plz help
Thanks a lot for all the help...

**Comment by jarrpa** [comment]
I'm not sure what version/commit of gk-deploy you're running, but I'll paste the prompt text of all the requirements here just to be sure. Note the kernel modules, in particular:
```
The client machine that will run this script must have:
 * Administrative access to an existing Kubernetes or OpenShift cluster
 * Access to a python interpreter 'python'
Each of the nodes that will host GlusterFS must also have appropriate firewall
rules for the required GlusterFS ports:
 * 2222  - sshd (if running GlusterFS in a pod)
 * 24007 - GlusterFS Management
 * 24008 - GlusterFS RDMA
 * 49152 to 49251 - Each brick for every volume on the host requires its own
   port. For every new brick, one new port will be used starting at 49152. We
   recommend a default range of 49152-49251 on each host, though you can adjust
   this to fit your needs.
The following kernel modules must be loaded:
 * dm_snapshot
 * dm_mirror
 * dm_thin_pool
For systems with SELinux, the following settings need to be considered:
 * virt_sandbox_use_fusefs should be enabled on each node to allow writing to
   remote GlusterFS volumes
```

Note that this time, since the topology loaded and possibly formatted the storage devices, you will need to do wipe them as well as the Gluster configs and resources. I think `wipefs -a <device>` should do. Afterwards, verify that `vgs` and `pvs` show no VGs or PVs (respectively) on any of the target devices. And, of course, if you care about these systems BE VERY CAREFUL. :)

**Comment by panigrahis** [comment]
@jarrpa Thanks for the info.
I double checked them and the modules seem to be present. 

root@UbuntuMaster:/etc/kubernetes/Private-Load-Balancer/gluster-kubernetes/deploy# cat check_mod_loded.sh
#!/bin/sh

MODULE="$1"

if lsmod | grep "$MODULE" &> /dev/null ; then
  echo "$MODULE is loaded!"
  exit 0
else
  echo "$MODULE is not loaded!"
  exit 1
fi

root@UbuntuMaster:/etc/kubernetes/Private-Load-Balancer/gluster-kubernetes/deploy# ./check_mod_loded.sh dm_snapshot
dm_snapshot is loaded!
root@UbuntuMaster:/etc/kubernetes/Private-Load-Balancer/gluster-kubernetes/deploy# ./check_mod_loded.sh dm_mirror
dm_mirror is loaded!
root@UbuntuMaster:/etc/kubernetes/Private-Load-Balancer/gluster-kubernetes/deploy# ./check_mod_loded.sh dm_thin_pool
dm_thin_pool is loaded!

Not sure whats going wrong

**Comment by panigrahis** [comment]
Oops my script is wrong...
Shit man...

**Comment by panigrahis** [comment]
@jarrpa where can i get these modules... and any pointers as to how they can be loaded to kernel.
Never done this before and do not want to screw up again.

**Comment by jarrpa** [comment]
@panigrahis No problem. :) So does that mean you have a working GlusterFS cluster with the images I provided?

**Comment by panigrahis** [comment]
@jarrpa just completed the deployment
Waiting for heketi pod to start ... OK

heketi is now running and accessible via http://10.244.17.2:8080 . To run
administrative commands you can install 'heketi-cli' and use it as follows:

  # heketi-cli -s http://10.244.17.2:8080 --user admin --secret '<ADMIN_KEY>' cluster list

You can find it at https://github.com/heketi/heketi/releases . Alternatively,
use it from within the heketi pod:

  # /usr/local/bin/kubectl -n default exec -it <HEKETI_POD> -- heketi-cli -s http://localhost:8080 --user admin --secret '<ADMIN_KEY>' cluster list

For dynamic provisioning, create a StorageClass similar to this:

---
apiVersion: storage.k8s.io/v1beta1
kind: StorageClass
metadata:
  name: glusterfs-storage
provisioner: kubernetes.io/glusterfs
parameters:
  resturl: "http://10.244.17.2:8080"


Thanks for all the

**Comment by panigrahis** [comment]
Just got the deployment successfully done:

Waiting for heketi pod to start ... OK

heketi is now running and accessible via http://10.244.17.2:8080 . To run
administrative commands you can install 'heketi-cli' and use it as follows:

  # heketi-cli -s http://10.244.17.2:8080 --user admin --secret '<ADMIN_KEY>' cluster list

You can find it at https://github.com/heketi/heketi/releases . Alternatively,
use it from within the heketi pod:

  # /usr/local/bin/kubectl -n default exec -it <HEKETI_POD> -- heketi-cli -s http://localhost:8080 --user admin --secret '<ADMIN_KEY>' cluster list

For dynamic provisioning, create a StorageClass similar to this:

---
apiVersion: storage.k8s.io/v1beta1
kind: StorageClass
metadata:
  name: glusterfs-storage
provisioner: kubernetes.io/glusterfs
parameters:
  resturl: "http://10.244.17.2:8080"

**Comment by panigrahis** [comment]
The deployment was successful....... finally....
Thanks for all the help
The next step would be to create a dynamic storage provider.

Waiting for heketi pod to start ... OK

heketi is now running and accessible via http://10.244.17.2:8080 . To run
administrative commands you can install 'heketi-cli' and use it as follows:

**Comment by jarrpa** [comment]
@panigrahis Godd to hear! :D 

@inspired-geek OKAY SO after hijacking the thread for so long: Do you feel your initial issue is sufficiently resolved or at least worked around? If so, please feel free to close this Issue and I'll open up a new issue to specifically address the systemd-in-container-related problems. :)

**Comment by jarrpa** [comment]
Closing this issue. If @inspired-geek needs anything further, they can reopen it. :)

**Comment by devent** [comment]
@jarrpa 
Hi jarrpa. I'm having the same issue and I looked at your modified daemonset yaml file. But I have a question. Why do you mount `/etc/target`? I don't think it's used in the Docker container.

**Comment by jarrpa** [comment]
@devent It's used by an upcoming feature called gluster-block, which is included in the container builds I made.

**Comment by devent** [comment]
@jarrpa Thank you for the info. But is `/etc/target` used in any way for the issue discussed here?

**Comment by jarrpa** [comment]
@devent Only in the sense that it is required if you want to use the experimental container images. Otherwise it has no technical bearing on the core issue.

**Comment by devent** [comment]
@jarrpa 
Hi jarrpa. I just tried your dev image, but I get the error:

```
root@robobee-test:~# kubectl -n kube-system logs glusterfs-cppcg
/usr/sbin/gluster-setup.sh: line 167: /usr/sbin/rpcbind: No such file or directory
Killing processes ... OK
```

I looked into your `jarrpa/gluster-ubuntu-minimal:latest` and I see that the script is using `/usr/sbin/rpcbind` but in the Ubuntu image the tool is available in `/sbin/rpcbind`.

```
erwin@work ~ $ docker run -d jarrpa/gluster-ubuntu-minimal sleep 30m
erwin@work ~ $ docker exec -it elated_leavitt bash                  
root@e5cc08946ba0:/# ls /usr/sbin/rpcbind
ls: cannot access '/usr/sbin/rpcbind': No such file or directory
root@e5cc08946ba0:/# which rpcbind
/sbin/rpcbind
```

That is different from Fedora:

```
erwin@work ~ $ docker run -d jarrpa/gluster-fedora-minimal sleep 30m
[root@fb6f10305b2d /]# ls -l /usr/sbin/rpcbind
lrwxrwxrwx 1 root root 14 May 30 16:14 /usr/sbin/rpcbind -> ../bin/rpcbind
[root@fb6f10305b2d /]# ls -l /sbin/rpcbind
lrwxrwxrwx 1 root root 14 May 30 16:14 /sbin/rpcbind -> ../bin/rpcbind
```

**Comment by jarrpa** [comment]
@devent Do you think it would be sufficient to change both to `/sbin/rpcbind`? Also, could you verify the location of `rpcinfo` in Ubuntu? Sorry to ask this of you, I am rather swamped at the moment.

**Comment by devent** [comment]
@jarrpa 
Sorry for the delay. I had no time from work and I just used the Fedora based image. The Fedora based image works fine. I think I already answered both of your questions :) . Clearly, on Ubuntu and Fedora `rcpbind` is under `/sbin` and Fedora just added a link. I'm no expert anyway, and I will test your updated Ubuntu based image.

**Comment by jarrpa** [comment]
@devent I still don't know the location of `rpcinfo`. :) But I'll hopefully get some time next week to mess around with the containers again.

**Comment by panigrahis** [comment]
@jarrpa i am getting the exact rpcinfo error stated above with ubuntu... How can this be fixed?

**Comment by jarrpa** [comment]
@panigrahis The startup script in the ubuntu-minimal container needs to be updated to point to the right location of rpcbind and rpcinfo. I just haven't had time to get back to that. :(

**Comment by devent** [comment]
@panigrahis It's fixed in my fork. But my fork also updated GlusterFS to version 3.12 because of https://github.com/gluster/gluster-kubernetes/issues/334

**Comment by soumyapanigrahi-k8s** [comment]
@devent thanks for the help. I am running K8S 1.8.
So creating new docker image with the fixes for path rpcinfo and rpcbind worked for me.

**Comment by josdotso** [comment]
I'm hitting this issue on:

- Kubernetes v1.7.3
- Docker 17.05.0-ce

What can I do @jarrpa ?  This thread is kind of hard to follow.

**Comment by jarrpa** [comment]
@josdotso I haven't run across that version of Docker before, but the solution at present is to change either Kube or Docker versions. If Docker, it should be < 1.13, if Kube it should be something other than 1.7.x (e.g. 1.6 or 1.8).
